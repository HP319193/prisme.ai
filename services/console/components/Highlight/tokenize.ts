interface Token {
  highlight: boolean;
  text: string;
  start: number;
  end: number;
}

export function tokenize(from: string, search: string) {
  if (!search) {
    return [
      {
        highlight: false,
        text: from,
        start: 0,
        end: from.length,
      },
    ];
  }
  const matches = Array.from(
    `${from}`.toLowerCase().matchAll(new RegExp(`${search.toLowerCase()}`, 'g'))
  );

  const len = search.length;
  const highlights: Token[] = matches.map(({ index = 0 }) => ({
    highlight: true,
    text: from.substring(index, index + len),
    start: index,
    end: index + len,
  }));

  const output: Token[] = [];
  Array.from(from).forEach((char, index) => {
    const prev = output[output.length - 1];

    const start = prev ? prev.end : 0;

    const highlighted = highlights.find(({ start: s }) => start === s);
    if (highlighted) {
      output.push(highlighted);
      return;
    }
    const nextHighlighted = highlights.find(({ start: s }) => s > start);
    const end = nextHighlighted ? nextHighlighted.start : from.length;
    const text = from.substring(start, end);
    if (!text) return;
    output.push({
      highlight: false,
      text,
      start,
      end,
    });
  });

  return output;
}
