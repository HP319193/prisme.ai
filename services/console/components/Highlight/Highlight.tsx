import {
  cloneElement,
  DetailedReactHTMLElement,
  ReactNode,
  useMemo,
} from 'react';
import { tokenize } from './tokenize';

interface HighlightProps {
  children: string;
  highlight?: string;
  component?: ReactNode;
}

export const Highlight = ({
  children,
  highlight = '',
  component = <span className="font-bold"></span>,
}: HighlightProps) => {
  const parts = useMemo(() => tokenize(children, highlight), [
    children,
    highlight,
  ]);

  return (
    <>
      {parts.map(({ highlight, text, start, end }) =>
        highlight && component ? (
          cloneElement(
            component as DetailedReactHTMLElement<any, HTMLElement>,
            {
              key: `${start}-${end}`,
              children: text,
            }
          )
        ) : (
          <span key={`${start}-${end}`}>{text}</span>
        )
      )}
    </>
  );
};

export default Highlight;
