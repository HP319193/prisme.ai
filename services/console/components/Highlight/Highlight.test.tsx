import renderer from 'react-test-renderer';
import Highlight from './Highlight';

it('should render', () => {
  const root = renderer.create(<Highlight highlight="foo">Foo bar</Highlight>);
  expect(root.toJSON()).toMatchSnapshot();
});

it('should render with custom component', () => {
  const root = renderer.create(
    <Highlight highlight="e" component={<span className="custom-class" />}>
      Exemple
    </Highlight>
  );
  expect(root.toJSON()).toMatchSnapshot();
});

it('should render with nothing', () => {
  const root = renderer.create(<Highlight>foo</Highlight>);
  expect(root.toJSON()).toMatchSnapshot();
});
