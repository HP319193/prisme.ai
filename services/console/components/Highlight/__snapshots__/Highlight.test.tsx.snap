// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
Array [
  <span
    className="font-bold"
  >
    Foo
  </span>,
  <span>
     bar
  </span>,
]
`;

exports[`should render with custom component 1`] = `
Array [
  <span
    className="custom-class"
  >
    E
  </span>,
  <span>
    x
  </span>,
  <span
    className="custom-class"
  >
    e
  </span>,
  <span>
    mpl
  </span>,
  <span
    className="custom-class"
  >
    e
  </span>,
]
`;

exports[`should render with nothing 1`] = `
<span>
  foo
</span>
`;
